plutus-tx/src/PlutusTx/Plugin/IsString.hs

```diff
@@ -0,0 +1,40 @@
+{-# LANGUAGE KindSignatures #-}
+
+module PlutusTx.Plugin.IsString (
+  Utf8Encoded (..),
+  RawBytes (..),
+  LiteralEncoding (..),
+) where
+
+import Prelude
+
+import Data.Kind (Type)
+import Data.String (IsString (..))
+
+{- Note [IsString instances and UTF-8 encoded string literals]
+
+GHC Core encodes string literals in UTF-8 by default.
+Plugin handles GHC Core expressions representing such literals specially:
+in compile type it undoes the UTF-8 encoding when string literal is used to construct
+a value of type 'BuiltinByteString' and preserves the UTF-8 encoding to construct 'BuiltinString'.
+
+The 'LiteralEncoding' type helps to distinguish between these two cases.
+
+The newtypes 'Utf8Encoded' and 'RawBytes' are used to indirectly provide IsString instances
+for other newtypes that wrap 'BuiltinString' or 'BuiltinByteString' i.e. 'TokenName':
+  - '"Мой Токен" :: TokenName'
+  - 'runRawBytes "Мой Токен" :: TokenName'
+
+-}
+
+data LiteralEncoding = UndoUtf8 | PreserveUtf8
+
+newtype Utf8Encoded (a :: Type) = Utf8Encoded {runUtf8Encoded :: a}
+
+instance (IsString a) => IsString (Utf8Encoded a) where
+  fromString = Utf8Encoded . fromString
```

Comment:
This is problematic since GHC may or may not inline `fromString`. You don't want it to be inlined, otherwise the plugin will try to compile `Utf8Encoded . fromString` which will fail. You can do something similar to what `instance IsString BuiltinByteString` does.

Reply:
Would `Magic.noinline` stop GHC inlining?

Reply:
`noinline` turns off inlining at a particular callsite. If you want something to never be inlined, use `OPAQUE`.
However, don't use `OPAQUE` on class methods, since GHC may still specialize it. Do what `IsString BuiltinByteString` does - delegate it to a function.

Reply:
Thanks for the clarification, will fix it.